hpl-2.2/man/man3/HPL_pdlaprnt.3
.TH HPL_pdlaprnt 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_pdlaprnt \- Print a distributed matrix A.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_pdlaprnt(\fR
\fB\&const HPL_T_grid *\fR
\fI\&GRID\fR,
\fB\&const int\fR
\fI\&M\fR,
\fB\&const int\fR
\fI\&N\fR,
\fB\&const int\fR
\fI\&NB\fR,
\fB\&double *\fR
\fI\&A\fR,
\fB\&const int\fR
\fI\&LDA\fR,
\fB\&const int\fR
\fI\&IAROW\fR,
\fB\&const int\fR
\fI\&IACOL\fR,
\fB\&const char *\fR
\fI\&CMATNM\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_pdlaprnt\fR
prints  to  standard  error a distributed matrix A. The
local pieces of  A  are sent to the process of coordinates  (0,0)  in
the grid and then printed.
.SH ARGUMENTS
.TP 8
GRID    (local input)           const HPL_T_grid *
On entry,  GRID  points  to the data structure containing the
process grid information.
.TP 8
M       (global input)          const int
On entry,  M  specifies the number of rows of the coefficient
matrix A. M must be at least zero.
.TP 8
N       (global input)          const int
On  entry,   N   specifies  the  number  of  columns  of  the
coefficient matrix A. N must be at least zero.
.TP 8
NB      (global input)          const int
On entry,  NB specifies the blocking factor used to partition
and distribute the matrix. NB must be larger than one.
.TP 8
A       (local input)           double *
On entry,  A  points to an  array of dimension (LDA,LocQ(N)).
This array contains the coefficient matrix to be printed.
.TP 8
LDA     (local input)           const int
On entry, LDA specifies the leading dimension of the array A.
LDA must be at least max(1,LocP(M)).
.TP 8
IAROW   (global input)          const int
On entry,  IAROW  specifies the row process coordinate owning
the  first row of A.  IAROW  must be  larger than or equal to
zero and less than NPROW.
.TP 8
IACOL   (global input)          const int
On entry,  IACOL  specifies  the  column  process  coordinate
owning the  first column  of A. IACOL  must be larger than or
equal to zero and less than NPCOL.
.TP 8
CMATNM  (global input)          const char *
On entry, CMATNM is the name of the matrix to be printed.
.SH SEE ALSO
.BR HPL_fprintf \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pauxil/HPL_pdlaprnt.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
void HPL_pdlaprnt
(
   const HPL_T_grid *               GRID,
   const int                        M,
   const int                        N,
   const int                        NB,
   double *                         A,
   const int                        LDA,
   const int                        IAROW,
   const int                        IACOL,
   const char *                     CMATNM
)
#else
void HPL_pdlaprnt
( GRID, M, N, NB, A, LDA, IAROW, IACOL, CMATNM )
   const HPL_T_grid *               GRID;
   const int                        M;
   const int                        N;
   const int                        NB;
   double *                         A;
   const int                        LDA;
   const int                        IAROW;
   const int                        IACOL;
   const char *                     CMATNM;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_pdlaprnt prints  to  standard  error a distributed matrix A. The
 * local pieces of  A  are sent to the process of coordinates  (0,0)  in
 * the grid and then printed.
 *
 * Arguments
 * =========
 *
 * GRID    (local input)                 const HPL_T_grid *
 *         On entry,  GRID  points  to the data structure containing the
 *         process grid information.
 *
 * M       (global input)                const int
 *         On entry,  M  specifies the number of rows of the coefficient
 *         matrix A. M must be at least zero.
 *
 * N       (global input)                const int
 *         On  entry,   N   specifies  the  number  of  columns  of  the
 *         coefficient matrix A. N must be at least zero.
 *
 * NB      (global input)                const int
 *         On entry,  NB specifies the blocking factor used to partition
 *         and distribute the matrix. NB must be larger than one.
 *
 * A       (local input)                 double *
 *         On entry,  A  points to an  array of dimension (LDA,LocQ(N)).
 *         This array contains the coefficient matrix to be printed.
 *
 * LDA     (local input)                 const int
 *         On entry, LDA specifies the leading dimension of the array A.
 *         LDA must be at least max(1,LocP(M)).
 *
 * IAROW   (global input)                const int
 *         On entry,  IAROW  specifies the row process coordinate owning
 *         the  first row of A.  IAROW  must be  larger than or equal to
 *         zero and less than NPROW.
 *
 * IACOL   (global input)                const int
 *         On entry,  IACOL  specifies  the  column  process  coordinate
 *         owning the  first column  of A. IACOL  must be larger than or
 *         equal to zero and less than NPCOL.
 *
 * CMATNM  (global input)                const char *
 *         On entry, CMATNM is the name of the matrix to be printed.
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   MPI_Comm                   Acomm;
   double                     * buf = NULL;
   int                        h, i, ib, icurcol=IACOL, icurrow=IAROW,
                              ii=0, j, jb, jj=0, mycol, myrow, npcol,
                              nprow, src;
/* ..
 * .. Executable Statements ..
 */
   (void) HPL_grid_info( GRID, &nprow, &npcol, &myrow, &mycol );
   Acomm = GRID->all_comm; 
   if( ( myrow == 0 ) && ( mycol == 0 ) )
      buf = (double*)malloc( (size_t)(NB) * sizeof( double ) );

   for( j = 0; j < N; j += NB )
   {
      jb = N-j; jb = Mmin( jb, NB );
      for( h = 0; h < jb; h++ )
      {
         (void) HPL_barrier( Acomm );

         for( i = 0; i < M; i += NB )
         {
            ib = M-i; ib = Mmin( ib, NB );
            if( ( icurrow == 0 ) && ( icurcol == 0 ) )
            {
               if( ( myrow == 0 ) && ( mycol == 0 ) )
                  HPL_dlaprnt( ib, 1, Mptr( A, ii, jj+h, LDA ), i+1,
                               j+h+1, LDA, CMATNM );
            }
            else
            {
               if( ( myrow == icurrow ) && ( mycol == icurcol ) )
               {
                  (void) HPL_send( Mptr( A, ii, jj+h, LDA ), ib, 0,
                                   9000+(j+h)*M+i, Acomm );
               }
               else if( ( myrow == 0 ) && ( mycol == 0 ) )
               {
                  src = HPL_pnum( GRID, icurrow, icurcol );
                  (void) HPL_recv( buf, ib, src, 9000+(j+h)*M+i,
                                   Acomm );
                  HPL_dlaprnt( ib, 1, buf, i+1, j+h+1, NB, CMATNM );
               }
            }
            if( myrow == icurrow ) ii += ib;
            icurrow = MModAdd1( icurrow, nprow );
            (void) HPL_barrier( Acomm );
         }
         ii = 0; icurrow = IAROW;
      }
      if( mycol == icurcol ) jj += jb;
      icurcol = MModAdd1( icurcol, npcol );
      (void) HPL_barrier( Acomm );
   }
   if( ( myrow == 0 ) && ( mycol == 0 ) && ( buf ) ) free( buf );
/*
 * End of HPL_pdlaprnt
 */
}
